### PR TITLE
Remote Auth Handler calls implemented

### PR DESCRIPTION
**Pull request by DomBlack**
This PR adds the ability for an Encore service (or gateway) to make a remote call to another container to run an auth handler.

This is required when the auth handler requires access to infrastructure that only the service it is defined in has access to. So all other containers will route their auth calls to that container.

This PR also fixes an issue where secrets would cause a panic in services when not set, even if those services where not running within the container.

### CODE REVIEW

File: runtime/appruntime/apisdk/api/auth_remote.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+	// Create the auth request
+	authReq, err := http.NewRequestWithContext(c.ctx, http.MethodPost, r.authURL, nil)
+	if err != nil {
+		r.logger.Err(err).Msg("unable to create auth request")
+		return model.AuthInfo{}, errs.Wrap(err, "unable to create auth request")
+	}
+
+	// Copy over any data which might be needed by the auth handler (query string, headers and cookies)
+	authReq.URL.RawQuery = c.req.URL.RawQuery
+	for k, v := range c.req.Header {
```
Comment (eandre) [inline runtime/appruntime/apisdk/api/auth_remote.go:50]: Probably we could just `maps.Clone` followed by deleting the `X-Encore-Auth` header

File: runtime/appruntime/apisdk/api/auth_remote.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+			continue
+		}
+
+		for _, vv := range v {
+			authReq.Header.Add(k, vv)
+		}
+	}
+
+	// Add call meta data to the request, so the receiving service can use to allow access to the auth handler
+	// This also passes a shared TraceID to the auth handler, so it and the microservice called after the gateway
```
Comment (eandre) [inline runtime/appruntime/apisdk/api/auth_remote.go:62]: Nit: we don't typically specify "microservice" but just "service"
```suggestion
	// This also passes a shared TraceID to the auth handler, so it and the service called after the gateway
```

File: runtime/appruntime/apisdk/api/encore_routes.go [hunk 3]
```diff
@@ -12,6 +12,7 @@ import (
 func (s *Server) registerEncoreRoutes() {
 	s.encore.HandlerFunc(wildcardMethod, "/healthz", s.handleHealthz)
 	s.encore.Handle("POST", "/pubsub/push/:subscription_id", s.handlePubsubPush)
+	s.encore.Handle("POST", "/auth_handler", s.handleRemoteAuthCall)
```
Comment (eandre) [inline runtime/appruntime/apisdk/api/encore_routes.go:15]: Can we change the URL to `/authhandler` instead? Underscores in URLs are a bit unusual I think

File: v2/codegen/infragen/secretsgen/secretsgen.go [hunk 4]
```diff
@@ -27,12 +29,21 @@ func Gen(gen *codegen.Generator, pkg *pkginfo.Package, secrets []*secrets.Secret
...
 		}
 
+		svcName := strconv.Quote(
+			option.
+				Map(
+					svc,
+					func(svc *app.Service) string { return svc.Name },
+				).
+				GetOrElse(""),
+		)
```
Comment (eandre) [inline v2/codegen/infragen/secretsgen/secretsgen.go:39]: Can we make this like two lines? If we move the closure to a separate variable the rest should fit in one line

File: runtime/appruntime/apisdk/api/gateway.go [hunk 5]
```diff
@@ -36,18 +38,58 @@ func (s *Server) createGatewayHandlerAdapter(h Handler) httprouter.Handle {
...
+		meta := CallMetaFromContext(req.Context())
+
+		info, proceed := s.runAuthHandler(h, s.NewIncomingContext(w, req, toUnnamedParams(ps), meta))
+		if proceed {
+			meta.Internal = &InternalCallMeta{
+				Caller: GatewayCaller{
+					ServiceName: h.ServiceName(),
+					Endpoint:    h.EndpointName(),
+				},
+				AuthUID:  string(info.UID),
```
Comment (eandre) [inline runtime/appruntime/apisdk/api/gateway.go:55]: Are we guaranteed that `info` is non-nil if `proceed` is true? (I forget the semantics of `runAuthHandler`)

Comment (DomBlack) [reply runtime/appruntime/apisdk/api/gateway.go:55]: `info` isn't a pointer - thus we always garanutee at least a zero value info struct here.

**Review by eandre — APPROVED** [review]

**Pull request merged**